### ISSUE TITLE
De-parenting has unwanted side effects

### ISSUE THREAD

**Issue by rogerl50** [ArnoldSmith86/virtualtabletop#12]
If I'm editing a widget's JSON and remove its parent attribute, or add one, then usually it disappears off the board, or at least moves to some unexpected location. This of course is because the coordinates change from absolute to relative. It would be really user-friendly if a change in parentage caused the code to revisit the coordinates, probably by assuming that the widget is at the place on the screen where the user wants it and computing new xy coordinates from the position of the holder.

**Comment by UltimateGeek** [comment]
Another scenario of bad coordinates I've run into is:
1. I hand modify a widget to make it a child of another widge
2. I forget to modify the X & Y values to be relative
3. It disappears off the board
4. So I have to move the parent's position on the board to recover the child widget

**Comment by rogerl50** [comment]
Yes. And sometimes you don't know where it went, and sometimes it's off the board and you have to go to Ghetto to find it. That's basically one of the two scenarios I was trying to describe in the original issue.

**Comment by RaphaelAlvez** [comment]
This can be easily fixed using a macro on Ghetto. I was thinking about creating on for parenting but I may do both

**Comment by robartsd** [comment]
Perhaps the built in JSON editor should _offer_ to change coordinates if parent has changed but x, y position has not.

**Comment by 96LawDawg** [comment]
Closed by #521.

**Comment by rogerl50** [comment]
This is not the same problem. While `fixedParent` prevents de-parenting when moving a widget, this PR wants the widget to not physically move onscreen when de-parented in the editor.

**Comment by 96LawDawg** [comment]
Sure.  Makes sense. I was too eager to close some old issues. Actually this might be easier to solve now with _absoluteX and Y.

**Comment by robartsd** [comment]
With #903 setting `x` to `_absoluteX`, `y` to `_absoluteY`, `scale` to `_absoluteScale`, and `rotation` to `_absoluteRotation`, then setting `parent` to `null` should cause the widget to not move at all except for any layering effect changes.

**Comment by robartsd** [comment]
One difficulty with fixing this is that the JSON editor tries to update live as you type (which generally we want to keep), so there is not really an way to determine if x and/or y have been changed without a fairly complicated system for tracking edits.

We could add a Change Parent button. I think the button should give the option to update x,y so that the widget's relative position in the room does not change (default option if widget parent get changed to or from null, but defaulting to keeping the widget x, y unchanged if changing from one parent to another).

**Comment by 96LawDawg** [comment]
I've been thinking about this issue as well.  I don't think we need or should do that.  It is overly complicated.  When building a room, I sometimes change parents just by changing the parent name by one letter or number.  An automated process would get in the way.

And after more thought, I don't think this is as big as a problem as it once was.  We have a zoomed out room view, we have the JSON Editor with its search features and the tree.  None of that existed when this issue was created.  The fact that it is 2 years old with no solution has as much to do with the fact that we have adequate work around solutions as it does with the fact that it is a little complicated to fix.

I would close this issue.

**Comment by 96LawDawg** [comment]
Closed by #2884. If using the JSON editor, deleting or adding a parent will still cause the widget to move to the new room coordinates. However, if using the new Edit Widgets UI, this does not happen.  A good compromise, I think, especially since people using the new UI would likely be more surprised by a widget jumping around when removing/adding a parent.
